kong/hooks.lua

```diff
@@ -0,0 +1,75 @@
+local _M = {}
+
+
+local hooks = {}
+
+
+local ipairs = ipairs
+local pack = table.pack
+local unpack = table.unpack
+local insert = table.insert
+
+
+local function wrap_hook(f)
+   return function(acc, ...)
+      if acc and not acc[1] then
+         return acc
+      end
+      return pack(f(...))
+   end
+end
+
+
+function _M.register_hook(name, hook, opts)
+  assert(type(hook) == "function", "hook must be a function")
+
+  hooks[name] = hooks[name] or {}
+
+  local f
+  if opts and opts.raw then
+    f = hook
+  else
+    f = wrap_hook(hook)
+  end
+
+  insert(hooks[name], f)
+end
+
+
+-- Register a hook function.
```

Comment:
This whole doc comment belongs to `_M.register_hook` (it is documenting the wrong function).